### PR TITLE
[tune] Only sync down from cloud if needed

### LINKED ISSUE #26714: Ray component: Tune S3 sync failure on restore_from_object
### What happened + What you expected to happen

In ray tune 1.13 I am facing an error I had not previously seen in 1.12. This error prevents tune from loading from checkpoints when using e.g. PBT 

The error is as follows when using SyncConfig in tune.run with an pre existing s3 bucket
```

2022-07-18 19:54:01,627 ERROR worker.py:95 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): ray::TFTrainble.restore_from_object() (pid=15654, ip=172.31.14.113, repr=<ray_tf_tuner.TFTrainble object at 0x7ff2d3aa0fd0>)
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/tune/trainable.py", line 606, in restore_from_object
    self.restore(checkpoint_path)
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/tune/trainable.py", line 553, in restore
    sleep_time=1,
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/tune/utils/util.py", line 132, in retry_fn
    fn()
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/tune/trainable.py", line 550, in <lambda>
    lambda: checkpoint.to_directory(local_dir),
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/ml/checkpoint.py", line 332, in to_directory
    download_from_uri(uri=external_path, local_path=path)
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/ml/utils/remote_storage.py", line 159, in download_from_uri
    pyarrow.fs.copy_files(bucket_path, local_path, source_filesystem=fs)
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/pyarrow/fs.py", line 250, in copy_files
    chunk_size, use_threads)
  File "pyarrow/_fs.pyx", line 1216, in pyarrow._fs._copy_files
  File "pyarrow/error.pxi", line 114, in pyarrow.lib.check_status
OSError: Path does not exist 'rohit-annigeri/test/experiments/results/tf_tune/TFTrainble_d526b_00000_0_loss=FocalLoss,lr=0.0010,optimizer=Adam_2022-07-18_19-50-12/tmpl0slh7j6restore_from_object'
```

Using the reproduction script below should see the same error 
```
2022-07-19 14:59:46,664	ERROR worker.py:95 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): ray::PBTBenchmarkExample.restore_from_object() (pid=17226, ip=172.31.14.45, repr=<ray-test.PBTBenchmarkExample object at 0x7ffa78261950>)
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/tune/trainable.py", line 606, in restore_from_object
    self.restore(checkpoint_path)
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/tune/trainable.py", line 553, in restore
    sleep_time=1,
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/tune/utils/util.py", line 132, in retry_fn
    fn()
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/tune/trainable.py", line 550, in <lambda>
    lambda: checkpoint.to_directory(local_dir),
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/ml/checkpoint.py", line 332, in to_directory
    download_from_uri(uri=external_path, local_path=path)
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/ml/utils/remote_storage.py", line 159, in download_from_uri
    pyarrow.fs.copy_files(bucket_path, local_path, source_filesystem=fs)
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/pyarrow/fs.py", line 250, in copy_files
    chunk_size, use_threads)
  File "pyarrow/_fs.pyx", line 1216, in pyarrow._fs._copy_files
  File "pyarrow/error.pxi", line 114, in pyarrow.lib.check_status
OSError: Path does not exist 'rohit-annigeri/ray-test/experiments/pbt_test/PBTBenchmarkExample_68c8f_00001_1_2022-07-19_14-59-42/tmp1kgdm6wkrestore_from_object'
```

After some digging around I found differences in storage client creation in Trainable class implementations between 1.12 and 1.13 (file path : ray/python/ray/tune/trainable.py )
specificallly 
in 1.12.0 
```

if self.uses_cloud_checkpointing:
    self.storage_client = self._create_storage_client()
```

in 1.13.0 and 1.13.1

```
if self.uses_cloud_checkpointing and self.sync_function_tpl:
    # Keep this only for custom sync functions and
    # backwards compatibility.
    # Todo (krfricke): We should find a way to register custom
    # syncers in Checkpoints rather than passing storage clients
    self.storage_client = self._create_storage_client()
```

since self.sync_function_tpl is None by default the and condition causes storage client not to created, which is different operation from 1.12.0. Simply changing it to or fixes this issue.

### Versions / Dependencies

ray==1.13.0 
python==3.7

### Reproduction script

```
#!/usr/bin/env python

import numpy as np
import argparse
import random
import os
import json

import ray
from ray import tune
from ray.tune.schedulers import PopulationBasedTraining


class PBTBenchmarkExample(tune.Trainable):
    def setup(self, config):
        self.lr = config["lr"]
        self.accuracy = 0.0  # end = 1000

    def step(self):
        midpoint = 100  # lr starts decreasing after acc > midpoint
        q_tolerance = 3  # penalize exceeding lr by more than this multiple
        noise_level = 2  # add gaussian noise to the acc increase

        if self.accuracy < midpoint:
            optimal_lr = 0.01 * self.accuracy / midpoint
        else:
            optimal_lr = 0.01 - 0.01 * (self.accuracy - midpoint) / midpoint
        optimal_lr = min(0.01, max(0.001, optimal_lr))

        # compute accuracy increase
        q_err = max(self.lr, optimal_lr) / min(self.lr, optimal_lr)
        if q_err < q_tolerance:
            self.accuracy += (1.0 / q_err) * random.random()
        elif self.lr > optimal_lr:
            self.accuracy -= (q_err - q_tolerance) * random.random()
        self.accuracy += noise_level * np.random.normal()
        self.accuracy = max(0, self.accuracy)

        return {
            "mean_accuracy": self.accuracy,
            "cur_lr": self.lr,
            "optimal_lr": optimal_lr,  # for debugging
            "q_err": q_err,  # for debugging
            "done": self.accuracy > midpoint * 2,
        }

    def save_checkpoint(self, checkpoint_dir):
        data = {
            "accuracy": self.accuracy,
            "lr": self.lr,
        }
        print(data)
        ckpt_path = os.path.join(checkpoint_dir, 'ckpt.json')
        with open(ckpt_path, 'w') as fp:
            json.dump(data, fp)
        return checkpoint_dir

    def load_checkpoint(self, checkpoint_dir):
        ckpt_path = os.path.join(checkpoint_dir, 'ckpt.json')
        with open(ckpt_path, 'r') as fp:
            checkpoint = json.loads(fp.read())
        self.accuracy = checkpoint["accuracy"]

    def reset_config(self, new_config):
        self.lr = new_config["lr"]
        self.config = new_config
        return True


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--smoke-test", action="store_true", help="Finish quickly for testing"
    )
    parser.add_argument(
        "--cluster", action="store_true", help="Distribute tuning on a cluster"
    )
    parser.add_argument(
        "--server-address",
        type=str,
        default=None,
        required=False,
        help="The address of server to connect to if using Ray Client.",
    )
    args, _ = parser.parse_known_args()

    if args.server_address:
        ray.init(f"ray://{args.server_address}")
    elif args.cluster:
        ray.init(address="auto")
    elif args.smoke_test:
        ray.init(num_cpus=2)  # force pausing to happen for test
    else:
        ray.init()

    pbt = PopulationBasedTraining(
        time_attr="training_iteration",
        perturbation_interval=20,
        hyperparam_mutations={
            # distribution for resampling
            "lr": lambda: random.uniform(0.0001, 0.02),
            # allow perturbations within this set of categorical values
            "some_other_factor": [1, 2],
        },
    )

    analysis = tune.run(
        PBTBenchmarkExample,
        name="pbt_test",
        scheduler=pbt,
        metric="mean_accuracy",
        mode="max",
        reuse_actors=True,
        checkpoint_freq=10,
        verbose=True,
        stop={
            "training_iteration": 200,
        },
        num_samples=8,
        config={
            "lr": 0.0001,
            "some_other_factor": 1,
        },
        resume=False,
        sync_config=tune.SyncConfig(
            # use your own s3 path
            upload_dir="s3://rohit-annigeri/ray-test/experiments/"
        )
    )

    print("Best hyperparameters found were: ", analysis.best_config)
```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by krfricke**
Signed-off-by: Kai Fricke <kai@anyscale.com>

## Why are these changes needed?

Currently, trainables will try to sync up/down temporary checkpoints from cloud storage, leading to errors. These erros come up e.g. with PBT, which heavily uses saving/restoring from objects.

Instead, we should not sync these temporary checkpoints up at all, and we should generally not sync down if a local checkpoint directory exists, which will prevent us also from trying to sync down non-existent temporary checkpoint directories.

See #26714

## Related issue number

Closes #26714

Related PR for 1.13.1: #26717

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/trainable/trainable.py [hunk 1]
```diff
@@ -487,7 +490,8 @@ def save(self, checkpoint_dir: Optional[str] = None) -> str:
         TrainableUtil.write_metadata(checkpoint_dir, metadata)
 
         # Maybe sync to cloud
-        self._maybe_save_to_cloud(checkpoint_dir)
+        if not prevent_upload:
```
Comment (xwjiang2010) [inline python/ray/tune/trainable/trainable.py:493]: How about syncing to driver?
I recall seeing these tmp folders on driver...

Comment (krfricke) [reply python/ray/tune/trainable/trainable.py:493]: As discussed, out of scope for this PR :-)

Comment (krfricke) [reply python/ray/tune/trainable/trainable.py:493]: (I'll address that in a different PR - this PR actually makes PBT fail while syncing to driver is not great but at least the experiment will continue to run).

**Review by xwjiang2010 — APPROVED** [review]

**Comment by xwjiang2010** [comment]
let’s revisit this whole pbt/syncing business after the release push. Ideally I would like to also add a e2e test using the user's repro script, if it makes sense for us.

**Pull request merged**